Fix failure to write '#' to key bindings file

svobs opened the pull request:
- [X] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [X] This implements/fixes issue #4267.

---

**Description:**

- `#` and ` ` could previously be entered into the Key Bindings editor by the user, but they would get written out to the `.conf` file as-is, so when they were read in again on the next launch, they would be ignored. Even more confusing for the user, the values in memory would still work, but only until the next launch.
- Since there is no other legitimate reason for a user to enter these into the editor, this patch alters the logic so that `#` is written out as `SHARP` and ` ` is written out as `SPACE`.

svobs committed:
Fix '#' and ' ' handling in Key Bindings editor by always escaping them when writing to file.

low-batt commented on iina/PrefKeyBindingViewController.swift:
```diff
@@ -325,11 +326,14 @@ class PrefKeyBindingViewController: NSViewController, PreferenceWindowEmbeddable
+    let keyMappings = mappingController.arrangedObjects as! [KeyMapping]
+    for mapping in keyMappings {
+      mapping.rawKey = KeyCodeHelper.escapeReservedMpvKeys(mapping.rawKey)
+    }
```
No need to change, but could use `forEach` instead of `for`:
```swift
keyMappings.forEach { $0.rawKey = KeyCodeHelper.escapeReservedMpvKeys($0.rawKey) }

```

low-batt replied:
Almost worked for me.

With these changes the `.conf` file now contains:
```text
SHARP quit
SPACE quit
```
The key bindings all look correct in settings. Looks as expected when I toggle `Display raw values`. When I press space IINA quits. **BUT** when I press # IINA responds with the unrecognized key sound and does not quit.

svobs replied:
> No need to change, but could use `forEach` instead of `for`:
> 
> ```swift
> keyMappings.forEach { $0.rawKey = KeyCodeHelper.escapeReservedMpvKeys($0.rawKey) }
> ```

I actually prefer to use a traditional `for` loop instead of a `forEach`, unless there's a really good reason to use it. I like having named variables and being able to introduce `break`, and maybe there's some jadedness from using debuggers which can't handle closures as well and make it harder to debug. I get the sense that some developers tell themselves they're making their code more efficient by reducing the number of lines written, but I think usually they're just increasing the bug density.

> BUT when I press # IINA responds with the unrecognized key sound and does not quit.

Well...yikes. I dug into this and found a whole bunch of bugs.... will comment more below.

low-batt replied:
Yes, that is why I mentioned no need to change. I definitely will use a traditional `for` if the situation makes it useful to be able to use `break` or `continue`.

svobs committed:
Fix bugs in mpv key binding normalization which could cause '#' and '+' keys to be ignored, and caused special keys which have "uppercase" versions (e.g. '4'->'$') to be ignored if using an explicit "Shift+" in the .conf file. Also fix a bug which caused equalsIgnoreCase to do a case sensitive comparison.

svobs commented:
Well...when I dug into your feedback, I found a bunch of issues with parsing. I suppose I could pack these changes into a separate PR...

Background refresher:

- There are multiple ways to represent a key sequence in the spec, but IINA does a string comparison to match the user's typed key(s) to a given key binding. Thus, I added a bunch of code to "normalize" it into a common format when doing the comparison (a lot of which used a bunch of code that was already there in `KeyCodeHelper` but tied it together).

1: SHARP & PLUS

There were several problems with the code which deals with special characters. The fix which allowed the `SHARP` key to be written to file broke the code which recognized whether that key was pressed. This underlying issue was that `SHARP` and `PLUS` are special token characters used by mpv, so they need to always been normalized to their whole-word representations instead of their literal `#` and `+` versions. But they weren't being normalized at all.

For these special symbols, the mpv `.conf` file spec allows for either literal (e.g., `+`)  or whole-word identifier (`PLUS`). When MacOS sends a key event to IINA, it is always translated to the literal version by default, so that's what it uses for the comparison by default. That means that anything specified in the whole-word version wouldn't be recognized when typed.

2: Uppercase handling was broken for non-alpha uppercase chars

There some code which I never dug into or tested, but with testing is clearly broken. The check for `key.lowercased() != key.uppercased()` doesn't do what we want to detect keys like `>` (we wanted it to detect it as a shifted `.` key). Likewise, the `event.readableKeyDescription` code is not returning valid data.

So I re-purposed the `keyMap` data structure in `KeyCodeHelper`, which largely wasn't being used and only needed a few fixes, to add a check for these special chars.

3: Case insensitive comparison broken

Most embarrassingly, `localizedCompare()` was being used instead of `localizedCaseInsensitiveCompare()`, which meant that `equalsIgnoreCase()` was actually not ignoring case... So this showed up when I tried `Shift+shift++` as a test case.

I put a few examples in the code comments which illustrate what was breaking. It sounds like a lot but most of these will probably only be noticed by people coming to IINA from mpv.

low-batt commented on iina/KeyCodeHelper.swift:
```diff
@@ -255,28 +267,40 @@ class KeyCodeHelper {
-      if rawKeyChar != char {
-        modifiers.remove(.shift)
-      }
+      // The char in `charactersIgnoringModifiers` will be either uppercase or lowercase,
```
From the Swift DocC section [Add a Basic Description for Each Symbol](https://www.swift.org/documentation/docc/writing-symbol-documentation-in-your-source-files#Add-a-Basic-Description-for-Each-Symbol):
> If a symbol already has a source comment that begins with two forward slashes (//), insert an additional forward slash (/) to convert it to a documentation comment.

I'm pretty sure [Markdown formatting](https://www.swift.org/documentation/docc/formatting-your-documentation-content) can only be used in swift documentation comments. I'm having trouble finding a reference, but I believe documentation comments will only be recognized before declarations, not inlined in a code block.

low-batt commented on iina/KeyCodeHelper.swift:
```diff
@@ -294,40 +318,58 @@ class KeyCodeHelper {
+  /// Ex04: "meta+shift+k" → "Meta+K"
+  /// Ex05: "esc" → "ESC"
+  /// Ex06: "CTRL+SHIFT+SHIFT+SHIFT+x" → "Ctrl+X"
+  /// _
```
In Xcode, clicking on `normalizeSingleMpvKeystroke` and then on the `?` button Xcode will show you what the rendered documentation looks like. 
```text
Normalizes a single “press” of possibly multiple keys (as joined with ‘+’) Examples: Ex01: “Shift+-” → “” Ex02: “Shift+=” → “PLUS” Ex03: “+-+” → “PLUS-PLUS” Ex04: “meta+shift+k” → “Meta+K” Ex05: “esc” → “ESC” Ex06: “CTRL+SHIFT+SHIFT+SHIFT+x” → “Ctrl+X” _
```

Markdown squishes it into one line. Maybe make use of a Markdown numbered list?

With a quick change to:
```swift
/// Normalizes a single "press" of possibly multiple keys (as joined with '+')
///
/// _Examples:_
/// 1. "Shift+-" → "_"
/// 2. "Shift+=" → "PLUS"
/// 3. "+-+" → "PLUS-PLUS"
/// 4. "meta+shift+k" → "Meta+K"
/// 5. "esc" → "ESC"
/// 6. "CTRL+SHIFT+SHIFT+SHIFT+x" → "Ctrl+X"
```

Xcode shows:

![issue-4267](https://user-images.githubusercontent.com/86170219/226032115-f624bfb1-e1b6-4b62-a3d5-e69d63807351.jpg)


Could still be improved. I have the [Basic Syntax](https://www.markdownguide.org/basic-syntax/) and [Extended Syntax](https://www.markdownguide.org/extended-syntax/)  pages from the [Markdown Guide](https://www.markdownguide.org/) site bookmarked.

There are various apps that support editing Markdown. I've been using the open source [Joplin](https://joplinapp.org/) application. Sources can be found on GitHub [here](https://github.com/laurent22/joplin).

low-batt commented:
Something it still not quite right.

As a test I created some bindings using the IINA UI that were stored like this by IINA in the `conf` file:
```text
SHARP quit
SPACE quit
+ quit
! quit
@ quit
$ quit
% quit
TAB quit
BS quit
ESC quit
{ quit
} quit
` quit
F2 quit
ENTER quit
```

A few of these bindings still did not work for me:

| Key | Results | Notes |
| --- |--- | --- |
| SHARP | FAILED | Made the unrecognized key sound |
| SPACE | Passed | |
| + | FAILED | Made the unrecognized key sound |
| ! | Passed | |
| @ | Passed | |
| $ | Passed | |
| % | Passed | |
| TAB | FAILED |**Did not** make the unrecognized key sound |
| BS | Passed | |
| ESC | Passed | |
| { | Passed | |
| } | Passed | |
| ` | Passed | |
| F2 | Passed | |
| ENTER | Passed | |

svobs committed:
Improve documentatiuon formatting. Fix broken key bindings. Note: TAB is still broken

svobs commented:
Thanks for the tips on XCode/Markdown formatting. It's one of those things I sort of care about but haven't spent any real time studying. I updated the comments in question so it looks nicer. 

On the use of caret symbols: even though I wasn't looking at the formatted version, I had fallen into using them as a practical way to indicate that the things inside identifiers and/or code snippets. It looks less ambiguous to me than using quotes or no punctuation at all. And XCode renders them in bold inside code blocks. Though I understand how some people might find them ugly or off-putting, so I'll try to avoid using them outside of declarations in the future.

I've been working in a few branches and I didn't notice the call to normalize the key input wasn't included (oops). Fixed in the latest commit.

But I need to look more at the issue of `TAB` not working. It's working in my big binding PR (#4160), but I made lots of fixes there which smooth out some weird stuff that is being done to the responder chain in the `develop` branch. Odd that the arrow keys work, but `TAB` doesn't...

low-batt reviewed (approved):
Pulled PR, built IINA, confirmed # key binding is working. Looks good to me.

svobs committed:
Merge branch 'iina:develop' into pr-fix-octothorpe-key

svobs committed:
Fix indentation of case statements

svobs commented:
Oops. I think I tried renaming some branches a couple of weeks ago, and this got automatically deleted when I did! Will see if this happened to any other active PRs.

uiryuu commented:
Thanks!
